### PR TITLE
API 4.4 changes

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
Attempt to incorporate API 4.4 changes.

**Stuff I'm not sure I handled it correctly:**
- `de_json` methods (`Chat` and `ChatPermission`)
- Deprecation warning for `all_members_are_administrators` in `Chat`
- new parameter sytax for `restrict_chat_members` in `Bot`
- Does something need to be changed for the new webhooks?
- Preparation for API 4.5, which brings nested `MessageEntities`: I adjusted `Message._parse_html` and `Message._parse_markdown` to allow for this. However, this only allows something like `_bold *within* italic_` and not `*bold _overlapping* italic_`. The announcement is not every clear, if that is also allowed …

Did I miss anything?

### CODE REVIEW

**Comment by Nukesor** [comment]
FYI: The 4.2 and 4.3 Changes are still not merged:
https://github.com/python-telegram-bot/python-telegram-bot/pull/1464

**Comment by Nukesor** [comment]
Pls remove the `WIP:` tag, if it's really ready for review :D.
Otherwise this might be overlooked.

**Comment by Bibo-Joshi** [comment]
Ah, overlooked that one. Thanks for the heads up :)

File: telegram/chatpermissions.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            page previews to their messages, implies :attr:`can_send_media_messages`.
+        can_change_info (:obj:`bool`, optional): True, if the user is allowed to change the chat
+            title, photo and other settings. Ignored in public supergroups.
+        can_invite_users (:obj:`bool`, optional): True, if the user is allowed to invite new users
+            to the chat.
+        can_pin_messages (:obj:`bool`, optional): True, if the user is allowed to pin messages.
+            Ignored in public supergroups.
+
+    Args:
+        can_send_messages (:obj:`bool`, optional): True, if the user is allowed to send text
```
Comment (Eldinnie) [inline telegram/chatpermissions.py]: I know it's annoying. But for parameters use:
``can_send_messages (:obj:`bool`, optional): True, if the user is allowed to send text messages, contacts, locations and venues.``
While for args use:
``can_send_messages (:obj:`bool`): Optional. True, if the user is allowed to send text messages, contacts, locations and venues.``

File: telegram/message.py [hunk 2]
```diff
@@ -978,33 +978,49 @@ def _parse_html(message_text, entities, urled=False):
         html_text = ''
         last_offset = 0
 
-        for entity, text in sorted(entities.items(), key=(lambda item: item[0].offset)):
-            text = escape(text)
```
Comment (Eldinnie) [inline telegram/message.py]: Untill we know how telegram handles nested entities, I'm not sure if we want to change this just yet.

File: tests/test_official.py [hunk 3]
```diff
@@ -121,6 +121,8 @@ def check_object(h4):
         ignored |= {'credentials'}
     elif name == 'PassportElementError':
         ignored |= {'message', 'type', 'source'}
+    elif name == 'Chat':
```
Comment (Eldinnie) [inline tests/test_official.py]: I think we can jsut remove this.

File: telegram/bot.py [hunk 4]
```diff
@@ -2720,24 +2716,15 @@ def restrict_chat_member(self, chat_id, user_id, until_date=None, can_send_messa
 
         Raises:
             :class:`telegram.TelegramError`
-
         """
         url = '{0}/restrictChatMember'.format(self.base_url)
 
-        data = {'chat_id': chat_id, 'user_id': user_id}
+        data = {'chat_id': chat_id, 'user_id': user_id, 'permissions': permissions.to_dict()}
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/bot.py]: `data.update(kwargs)` was there before (former line 2741, now line 2728).
Should I still add the `TODO:`?

File: telegram/chat.py [hunk 5]
```diff
@@ -106,10 +110,16 @@ def __init__(self,
         self.first_name = first_name
         self.last_name = last_name
         self.all_members_are_administrators = all_members_are_administrators
+        if all_members_are_administrators is not None:
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/chat.py]: Removed

File: telegram/chatpermissions.py (hunk 1 above)
Comment (Bibo-Joshi) [reply telegram/chatpermissions.py]: Yes, sorry. Changed.

File: tests/test_official.py (hunk 3 above)
Comment (Bibo-Joshi) [reply tests/test_official.py]: Removed

File: telegram/message.py (hunk 2 above)
Comment (Bibo-Joshi) [reply telegram/message.py]: I could reverse the corresponding commit and put it in a new branch, so we can get back to it when API 4.5 is released. Shall I?

File: telegram/bot.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Eldinnie) [reply telegram/bot.py]: woops, missed that. leave the todo out. can;t hurt to add kwargs

File: telegram/message.py (hunk 2 above)
Comment (Eldinnie) [reply telegram/message.py]: yes, that seems appropriate. It's good code if they handle it like this.

Comment (Bibo-Joshi) [reply telegram/message.py]: Done :) See #1508

**Pull request merged**